File: sdk/typescript/src/proofs/ai.ts
/**
 * AIProof helpers — assemble lightweight proof references from outputs.
 *
 * This module helps clients:
 *  1) Compute a canonical digest of an AI output (bytes / string / object)
 *  2) Package an AIProof "reference" object that nodes can verify against
 *     attestation bundles, trap receipts, and QoS metrics
 *  3) Produce/consume a minimal CBOR envelope for transport
 *
 * NOTE: This is a *client-side* helper. Consensus verification of the AI proof
 * (TEE attestation checks, traps, QoS) happens in node software. We keep types
 * broad to accommodate different providers and attestation formats.
 */

import { sha3_256 } from '../utils/hash'
import { bytesToHex, hexToBytes, isHex } from '../utils/bytes'
import { encodeCanonicalCBOR, decodeCBOR } from '../utils/cbor'

// ──────────────────────────────────────────────────────────────────────────────
// Types
// ──────────────────────────────────────────────────────────────────────────────

/** Generic attestation bundle (e.g., SGX/SEV/CCA) — provider-defined structure. */
export type AIAttestation = Record<string, unknown>

/** Optional trap receipts info used by verifiers for redundancy/trust checks. */
export interface AITrapInfo {
  /** Optional deterministic seed describing trap set selection (0x-hex). */
  trapSeed?: string
  /** Ratio of trap inputs that were correctly answered by provider [0..1]. */
  trapsRatio?: number
  /** Digest (0x-hex) of trap receipt set, if provided separately. */
  receiptsDigest?: string
}

/** Optional QoS/run-time metrics to assist pricing/SLA logic. */
export interface AIQoS {
  latencyMs?: number           // end-to-end latency for the job
  availability?: number        // long-term availability ratio [0..1]
  successRate?: number         // success ratio [0..1]
}

/** Canonical AI proof body expected by the chain's verifier. */
export interface AIProofBody {
  model: string                // identifier (e.g., gpt-4o-mini-2025-05)
  outputDigest: string         // 0x-hex sha3_256 digest of output payload
  outputMime?: string          // e.g., application/json, text/plain
  providerId?: string          // logical provider id (optional)
  jobId?: string               // queue/job id (optional but helps nullifier uniqueness)
  attestation: AIAttestation   // TEE/token bundle; opaque here
  traps?: AITrapInfo
  qos?: AIQoS
  // Future-extensible:
  [k: string]: unknown
}

/** Complete reference envelope used off-chain/on-wire. */
export interface AIProofRef {
  typeId: 'AIProof' | number
  body: AIProofBody
  /** Deterministic nullifier to prevent duplicated submissions. */
  nullifier: string            // 0x-hex
}

/** Builder inputs for creating an AIProofRef. */
export interface BuildAIProofInput {
  model: string
  /** Output payload (bytes | 0x-hex | UTF-8 string | JSON-like object). */
  output: Uint8Array | string | Record<string, unknown> | unknown[]
  outputMime?: string
  providerId?: string
  jobId?: string
  attestation: AIAttestation
  traps?: AITrapInfo
  qos?: AIQoS
}

/** Validation result (client-side sanity only). */
export interface ValidateResult {
  ok: boolean
  errors: string[]
  warnings: string[]
}

// ──────────────────────────────────────────────────────────────────────────────
// Public API
// ──────────────────────────────────────────────────────────────────────────────

/**
 * Compute a canonical digest for an AI output.
 *  - Uint8Array → bytes hashed as-is
 *  - 0x-hex string → parsed, hashed as bytes
 *  - other string → UTF-8 encoded then hashed
 *  - object/array → deterministically CBOR-encoded then hashed
 */
export function digestOutput(payload: BuildAIProofInput['output']): string {
  const bytes = toBytesCanonical(payload)
  return '0x' + bytesToHex(sha3_256(bytes))
}

/**
 * Build a minimal AIProofRef from the provided output + metadata.
 * The nullifier binds (providerId|jobId|model|outputDigest) under a domain tag.
 */
export function buildAIProofRef(input: BuildAIProofInput): AIProofRef {
  const outputDigest = digestOutput(input.output)
  const body: AIProofBody = {
    model: input.model,
    outputDigest,
    outputMime: input.outputMime,
    providerId: input.providerId,
    jobId: input.jobId,
    attestation: input.attestation,
    traps: normalizeTraps(input.traps),
    qos: normalizeQoS(input.qos)
  }
  const nullifier = computeNullifier(body)
  return { typeId: 'AIProof', body, nullifier }
}

/** Compute deterministic nullifier = H("ai.nullifier.v1"|providerId?|jobId?|model|outputDigest). */
export function computeNullifier(body: Pick<AIProofBody, 'providerId' | 'jobId' | 'model' | 'outputDigest'>): string {
  const dom = text('ai.nullifier.v1')

  const parts: Uint8Array[] = [dom]
  if (body.providerId) parts.push(text(body.providerId))
  if (body.jobId) parts.push(text(body.jobId))
  parts.push(text(body.model))
  parts.push(hexToBytes(normalizeHex(body.outputDigest)))

  return '0x' + bytesToHex(sha3_256(concat(...parts)))
}

/** Shallow client-side validation (format/ranges only). */
export function validateAIProofBody(body: AIProofBody): ValidateResult {
  const errors: string[] = []
  const warnings: string[] = []

  if (!body.model || typeof body.model !== 'string') errors.push('model missing/invalid')
  try { normalizeHex(body.outputDigest) } catch { errors.push('outputDigest must be 0x-hex sha3_256') }
  if (body.traps?.trapsRatio != null) {
    if (!(body.traps.trapsRatio >= 0 && body.traps.trapsRatio <= 1)) {
      errors.push('traps.trapsRatio must be in [0,1]')
    }
  }
  if (body.qos?.availability != null && !(body.qos.availability >= 0 && body.qos.availability <= 1)) {
    errors.push('qos.availability must be in [0,1]')
  }
  if (body.qos?.successRate != null && !(body.qos.successRate >= 0 && body.qos.successRate <= 1)) {
    errors.push('qos.successRate must be in [0,1]')
  }
  if (!body.attestation || typeof body.attestation !== 'object') {
    errors.push('attestation missing/invalid')
  }

  return { ok: errors.length === 0, errors, warnings }
}

/** Encode an AIProofRef into a minimal canonical CBOR envelope. */
export function toEnvelopeCBOR(ref: AIProofRef): Uint8Array {
  const env: AIProofRef = {
    typeId: typeof ref.typeId === 'number' ? ref.typeId : 'AIProof',
    body: {
      ...ref.body,
      outputDigest: normalizeHex(ref.body.outputDigest),
      traps: normalizeTraps(ref.body.traps),
      qos: normalizeQoS(ref.body.qos)
    },
    nullifier: normalizeHex(ref.nullifier)
  }
  return encodeCanonicalCBOR(env)
}

/** Decode a CBOR envelope back into a normalized AIProofRef. */
export function fromEnvelopeCBOR(data: Uint8Array): AIProofRef {
  const env = decodeCBOR(data) as AIProofRef
  if (!env || typeof env !== 'object') throw new Error('invalid envelope')
  if (!('typeId' in env) || !('body' in env) || !('nullifier' in env)) throw new Error('missing envelope fields')
  const out: AIProofRef = {
    typeId: (env.typeId === 'AIProof' || typeof env.typeId === 'number') ? env.typeId : 'AIProof',
    body: {
      ...env.body,
      outputDigest: normalizeHex(env.body.outputDigest),
      traps: normalizeTraps(env.body.traps),
      qos: normalizeQoS(env.body.qos)
    },
    nullifier: normalizeHex(env.nullifier)
  }
  return out
}

// ──────────────────────────────────────────────────────────────────────────────
// Internal helpers
// ──────────────────────────────────────────────────────────────────────────────

function normalizeHex(h: string): string {
  if (typeof h !== 'string') throw new Error('expected hex string')
  const s = h.startsWith('0x') || h.startsWith('0X') ? h : ('0x' + h)
  if (!isHex(s)) throw new Error(`invalid hex: ${h}`)
  return s.toLowerCase()
}

function text(s: string): Uint8Array {
  return new TextEncoder().encode(s)
}

function concat(...parts: Uint8Array[]): Uint8Array {
  if (parts.length === 0) return new Uint8Array(0)
  const n = parts.reduce((a, p) => a + p.length, 0)
  const out = new Uint8Array(n)
  let off = 0
  for (const p of parts) { out.set(p, off); off += p.length }
  return out
}

/** Normalize traps object & basic hex checks. */
function normalizeTraps(t?: AITrapInfo): AITrapInfo | undefined {
  if (!t) return undefined
  const traps: AITrapInfo = { ...t }
  if (traps.trapSeed != null) traps.trapSeed = normalizeHex(traps.trapSeed)
  if (traps.receiptsDigest != null) traps.receiptsDigest = normalizeHex(traps.receiptsDigest)
  return traps
}

/** Normalize QoS fields to numbers (drop NaN/undefined). */
function normalizeQoS(q?: AIQoS): AIQoS | undefined {
  if (!q) return undefined
  const out: AIQoS = {}
  if (q.latencyMs != null) out.latencyMs = Number(q.latencyMs)
  if (q.availability != null) out.availability = Number(q.availability)
  if (q.successRate != null) out.successRate = Number(q.successRate)
  return out
}

/** Canonical byte representation for digesting arbitrary "output" values. */
function toBytesCanonical(x: BuildAIProofInput['output']): Uint8Array {
  if (x instanceof Uint8Array) return x
  if (typeof x === 'string') {
    if (isHex(x)) return hexToBytes(x)
    return new TextEncoder().encode(x)
  }
  // For objects/arrays: canonical CBOR so all languages match the digest
  return encodeCanonicalCBOR(x as any)
}

// ──────────────────────────────────────────────────────────────────────────────
// Default export
// ──────────────────────────────────────────────────────────────────────────────

export default {
  digestOutput,
  buildAIProofRef,
  computeNullifier,
  validateAIProofBody,
  toEnvelopeCBOR,
  fromEnvelopeCBOR
}

File: sdk/rust/src/proofs/ai.rs
//! AIProof assembly helpers (developer-facing).
//!
//! This module helps SDK users assemble an **AI proof reference** from:
//! - Model output bytes (we hash to a digest bound in the proof)
//! - TEE attestation bundle from the provider
//! - Optional trap-circuit receipts (pass/total → ratio)
//! - Optional QoS metrics (latency, availability, success)
//!
//! It does **not** implement the full on-chain CBOR envelope; rather it
//! produces a forward-compatible JSON-serializable struct your app can
//! pass to node tooling that builds canonical envelopes per `proofs/`.

use crate::error::{Error, Result};
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use sha3::{Digest, Sha3_256};
use std::collections::BTreeMap;

/// TEE attestation bundle (opaque JSON; minimally tagged with `vendor`).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiAttestation {
    /// Provider/vendor marker, e.g. "intel-sgx", "amd-sev-snp", "arm-cca".
    pub vendor: String,
    /// Evidence bundle (e.g., quote/report/token JSON), provider-specific.
    pub evidence: JsonValue,
    /// Future-proof fields.
    #[serde(flatten)]
    pub extra: BTreeMap<String, JsonValue>,
}

/// Trap-circuit outcomes (used for correctness confidence).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiTraps {
    pub passed: u32,
    pub total: u32,
    /// Optional precomputed ratio; if absent, computed on validate().
    #[serde(default)]
    pub ratio: Option<f64>,
    #[serde(flatten)]
    pub extra: BTreeMap<String, JsonValue>,
}

impl AiTraps {
    pub fn compute_ratio(&self) -> f64 {
        if self.total == 0 {
            return 0.0;
        }
        (self.passed as f64) / (self.total as f64)
    }
}

/// QoS metrics observed for the job/provider.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct AiQos {
    /// Whole-request latency in milliseconds (provider reported or measured).
    #[serde(default)]
    pub latency_ms: Option<u64>,
    /// Recent availability estimate in [0.0, 1.0].
    #[serde(default)]
    pub availability: Option<f64>,
    /// Whether the request completed successfully.
    #[serde(default)]
    pub success: Option<bool>,
    /// Optional provider score in [0.0, 1.0] (SLA composite).
    #[serde(default)]
    pub score: Option<f64>,
    #[serde(flatten)]
    pub extra: BTreeMap<String, JsonValue>,
}

/// AI proof reference (what apps pass into proof builders).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiProofRef {
    /// SHA3-256 digest of the model output bytes (0x-hex, 32 bytes).
    pub output_digest: String,
    /// Size of the raw output in bytes (useful for pricing/quotas).
    pub output_size: u64,
    /// Optional MIME (e.g., "application/json", "text/plain", "image/png").
    #[serde(default)]
    pub mime: Option<String>,
    /// Redundant runs for the same prompt (>=1). Used for confidence boosting.
    #[serde(default)]
    pub redundancy: Option<u32>,
    /// Attestation bundle (TEE or equivalent).
    pub attestation: AiAttestation,
    /// Optional trap-circuit outcomes.
    #[serde(default)]
    pub traps: Option<AiTraps>,
    /// Optional QoS metrics.
    #[serde(default)]
    pub qos: Option<AiQos>,
    /// Free-form application metadata (kept out of consensus-critical fields).
    #[serde(default)]
    pub meta: BTreeMap<String, JsonValue>,
    /// Future-proof fields.
    #[serde(flatten)]
    pub extra: BTreeMap<String, JsonValue>,
}

impl AiProofRef {
    /// Construct from raw model `output` and an `attestation` bundle.
    /// - Computes `output_digest = 0x + sha3_256(output)`
    /// - Sets `output_size`
    pub fn from_output(
        output: &[u8],
        attestation: AiAttestation,
        traps: Option<AiTraps>,
        qos: Option<AiQos>,
    ) -> Self {
        let mut hasher = Sha3_256::new();
        hasher.update(output);
        let digest = hasher.finalize();
        Self {
            output_digest: hex0x(&digest),
            output_size: output.len() as u64,
            mime: None,
            redundancy: None,
            attestation,
            traps,
            qos,
            meta: BTreeMap::new(),
            extra: BTreeMap::new(),
        }
    }

    /// Validate basic invariants; fills derived fields (e.g., traps.ratio).
    pub fn validate(&mut self) -> Result<()> {
        // Digest should be 0x + 64 hex chars.
        if !self.output_digest.starts_with("0x") || self.output_digest.len() != 66 {
            return Err(Error::InvalidData(format!(
                "outputDigest must be 0x + 64 hex chars, got len={}",
                self.output_digest.len()
            )));
        }
        if let Some(r) = self.redundancy {
            if r == 0 {
                return Err(Error::InvalidData("redundancy must be >= 1".into()));
            }
        }
        // Trap ratio in [0,1].
        if let Some(t) = self.traps.as_mut() {
            let ratio = t.ratio.unwrap_or_else(|| t.compute_ratio());
            if !(0.0..=1.0).contains(&ratio) {
                return Err(Error::InvalidData(format!(
                    "trap ratio out of range: {ratio}"
                )));
            }
            t.ratio = Some(ratio);
        }
        // QoS bounds.
        if let Some(q) = self.qos.as_ref() {
            if let Some(a) = q.availability {
                if !(0.0..=1.0).contains(&a) {
                    return Err(Error::InvalidData(format!("availability out of range: {a}")));
                }
            }
            if let Some(s) = q.score {
                if !(0.0..=1.0).contains(&s) {
                    return Err(Error::InvalidData(format!("score out of range: {s}")));
                }
            }
        }
        Ok(())
    }
}

/// Build a ready-to-serialize AI proof reference with optional MIME and redundancy.
pub fn build_ai_proof_ref(
    output: &[u8],
    mime: Option<&str>,
    redundancy: Option<u32>,
    attestation: AiAttestation,
    traps: Option<AiTraps>,
    qos: Option<AiQos>,
) -> Result<AiProofRef> {
    let mut pref = AiProofRef::from_output(output, attestation, traps, qos);
    if let Some(m) = mime {
        pref.mime = Some(m.to_string());
    }
    if let Some(r) = redundancy {
        pref.redundancy = Some(r);
    }
    pref.validate()?;
    Ok(pref)
}

// ------------------------------- Utilities ----------------------------------

fn hex0x(bytes: impl AsRef<[u8]>) -> String {
    format!("0x{}", hex::encode(bytes))
}

// ---------------------------------- Tests -----------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn sample_attestation() -> AiAttestation {
        AiAttestation {
            vendor: "intel-sgx".into(),
            evidence: json!({
                "quote": "0xdeadbeef",
                "pckCertChain": ["-----BEGIN CERT-----..."]
            }),
            extra: BTreeMap::new(),
        }
    }

    #[test]
    fn build_and_validate_prefills_digest_and_size() {
        let output = b"{\"text\":\"hello\"}";
        let att = sample_attestation();
        let mut pref = AiProofRef::from_output(output, att, None, None);
        assert_eq!(pref.output_size, output.len() as u64);
        assert!(pref.output_digest.starts_with("0x"));
        assert_eq!(pref.output_digest.len(), 66);
        pref.validate().unwrap();
    }

    #[test]
    fn traps_ratio_is_computed_when_missing() {
        let output = b"ok";
        let att = sample_attestation();
        let traps = AiTraps {
            passed: 9,
            total: 10,
            ratio: None,
            extra: BTreeMap::new(),
        };
        let mut pref = AiProofRef::from_output(output, att, Some(traps), None);
        pref.validate().unwrap();
        let r = pref.traps.as_ref().unwrap().ratio.unwrap();
        assert!((r - 0.9).abs() < 1e-9);
    }

    #[test]
    fn qos_bounds_are_checked() {
        let output = b"ok";
        let att = sample_attestation();
        // Bad availability
        let qos = AiQos {
            availability: Some(1.5),
            ..Default::default()
        };
        let mut pref = AiProofRef::from_output(output, att, None, Some(qos));
        let err = pref.validate().unwrap_err();
        assert!(matches!(err, Error::InvalidData(_)));
    }

    #[test]
    fn full_builder_with_mime_and_redundancy() {
        let output = b"result";
        let att = sample_attestation();
        let pref = build_ai_proof_ref(
            output,
            Some("application/json"),
            Some(2),
            att,
            None,
            Some(AiQos {
                latency_ms: Some(120),
                availability: Some(0.999),
                success: Some(true),
                score: Some(0.98),
                extra: BTreeMap::new(),
            }),
        )
        .unwrap();
        assert_eq!(pref.mime.as_deref(), Some("application/json"));
        assert_eq!(pref.redundancy, Some(2));
    }
}
